### PR TITLE
feat: wrap safe ETH into WETH and add ERC-1271 signature validation

### PR DESCRIPTION
**Pull request by shinchann221**
Native ETH has no exit from a safe: `_processWithdrawal` encodes only `IERC20.transfer` and passes zero call values, so ETH sent to a safe is stranded.

**WETH wrapping**
- `receive()` on `EtherFiSafe` wraps incoming ETH into WETH. `WETH` is a constant — Optimism is the only live Cash stack, and the constructor rejects a deploy on any chain where that address has no code rather than shipping a safe that reverts on every incoming transfer.
- Passes ETH through untouched for three senders that need it native: WETH itself (unwrap would recurse, and carries only a 2300-gas stipend), an enabled module (Enso and BeHYPE pre-fund the safe then spend it as call value), and anything arriving mid-batch (OpenOcean reads its ETH output as a native balance delta).
- Mid-batch is tracked by a transient flag set in `execTransactionFromModule`, saved and restored so a nested batch cannot unset the outer one.
- `wrapEth()` is a permissionless sweep for balances already stranded, and no-ops mid-batch for the same reason.
- `TradingSafe` is unaffected — `receive()` is now `virtual` and only `EtherFiSafe` overrides it.

**ERC-1271**
- `isValidSignature` validates against the safe's existing owners and threshold, binding the hash twice: it must be the EIP-191 hash of a supplied `message`, and owners sign it wrapped in this safe's own EIP-712 domain.
- The preimage binding is what preserves the LendGateway invariant. EIP-712 digests begin `0x1901` and the EIP-191 prefix fixes byte 1 to `E`, so no message can hash to one — Permit2, Seaport and the Spoke's `setUserPositionManagersWithSig` stay unsatisfiable here even for a holder of threshold owner keys.
- `signature` is `abi.encode(bytes message, address[] signers, bytes[] signatures)`.

**Follow-ups**
- WETH must be in `whitelistedWithdrawAssets` on OP or wrapped ETH still cannot be withdrawn.
- Native-ETH module paths (`OpenOceanSwapModule` with `fromAsset == ETH`, the `EtherFiStakeModule` / `EtherFiLiquidModule` / `AaveV3Module` ETH branches, `CashLens.getMaxSourceable(ETH)`) will now see a zero balance; routing should quote WETH.
- `LendGateway`'s invariant comment still describes the pre-1271 state.

---

> [!NOTE]
> **High Risk**
> A single beacon upgrade repoints every OP safe at once, changing ETH handling and introducing ERC-1271 with security-critical hash binding; incorrect library linking or governance execution would affect all proxies.
> 
> **Overview**
> **`EtherFiSafe` on Optimism** now auto-wraps incoming ETH to WETH (with pass-throughs for WETH, low gas, module batches, enabled modules, and pauses) and adds permissionless **`wrapEth`** gated on **`dataProvider.paused()`**. **`EtherFiSafeCore`** tracks module batches with a transient flag so nested batches do not break ETH handling.
> 
> **ERC-1271** is implemented through a separately deployed **`SafeErc1271Lib`** (delegatecall) so bytecode stays under the size limit; validation only accepts EIP-191 message hashes, preserving the **LendGateway** invariant against foreign EIP-712 digests.
> 
> **Ops / rollout:** New deploy scripts pin the library link (`--libraries`), **`UpgradeSafeImpl`** is dev-only with prechecks, **`UpgradeSafeImplOP3CP`** emits 8h timelock Safe bundles plus fork simulation, and OP **`deployments.json`** records **`EtherFiSafeImpl`** / **`SafeErc1271Lib`** with pre-generated Gnosis JSON. **`ContractCodeChecker`** gains embedded-address checks for bytecode verification.
> 
> Tests cover WETH/`receive` edge cases, ERC-1271, and cash withdrawal of wrapped ETH.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 774d621bba27434ba173b745340338fbf7d52554. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/gnosis-txs/UpgradeSafeImplOP3CP.s.sol [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+            uint256 wethBefore = IERC20(OP_WETH).balanceOf(liveSafe);
+            address sender = address(uint160(uint256(keccak256("UpgradeSafeImpl.sim.sender"))));
+            vm.deal(sender, 1 ether);
+            vm.prank(sender);
+            (bool ok,) = liveSafe.call{ value: 1 ether }("");
+
+            require(ok, "SIM FAILED: live safe rejected an ETH transfer");
+            require(liveSafe.balance == 0, "SIM FAILED: ETH stayed native on a live safe");
+            require(IERC20(OP_WETH).balanceOf(liveSafe) == wethBefore + 1 ether, "SIM FAILED: ETH was not wrapped into WETH");
```
[the remark this answers is not in the document]
Comment (ethanai-ethfi) [reply scripts/gnosis-txs/UpgradeSafeImplOP3CP.s.sol]: Valid on live head 83883b9. The sim requires `liveSafe.balance == 0` after sending 1 ETH, but `receive()` only deposits `msg.value`; any native balance already on the safe remains. Smallest fix: capture `nativeBefore`, assert the transfer leaves `liveSafe.balance == nativeBefore` and WETH increases by 1 ETH, then call `wrapEth()` (or sweep first) to verify stranded ETH recovery.

[the remark this answers is not in the document]
Comment (shinchann221) [reply scripts/gnosis-txs/UpgradeSafeImplOP3CP.s.sol]: Valid, fixed in 2cde95b.

`receive` wraps only `msg.value`, so pre-existing stranded ETH stays native and `liveSafe.balance == 0` would fail on a correct upgrade — and the first-deployed safe is exactly the kind likely to hold some.

The assertion was also testing the wrong thing: a stranded balance is what this PR exists to clear, so the simulation now proves both mechanisms separately instead of assuming the safe started empty.

- capture `strandedBefore = liveSafe.balance`
- after the 1 ETH send: `balance == strandedBefore` and WETH `+1 ether` — proves arrival-time wrapping without touching the pre-existing balance
- then call `wrapEth()`: `balance == 0` and WETH `+1 ether + strandedBefore` — proves the sweep

File: scripts/UpgradeSafeImpl.s.sol [hunk 2]
```diff
@@ -1,35 +1,78 @@
...
+    ///      internal try/catch answer INVALID, which needs no owners and so works against a bare
+    ///      implementation. If the linked address held no code the DELEGATECALL would return nothing and
+    ///      decoding a bytes4 from it would revert, so this fails loudly either way.
+    function _assertErc1271LibraryIsLinked(address impl) internal view {
+        bytes memory message = "SafeErc1271Lib link check";
+        bytes32 hash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n25", message));
+        bytes memory blob = abi.encode(message, new address[](0), new bytes[](0));
+
+        require(EtherFiSafe(payable(impl)).isValidSignature(hash, blob) == bytes4(0xffffffff), "ERC-1271 path did not answer - SafeErc1271Lib is not linked to live code");
     }
```
[the remark this answers is not in the document]
Comment (ethanai-ethfi) [reply scripts/UpgradeSafeImpl.s.sol:77]: Partially valid on live head b41e852. Evidence: src/safe/EtherFiSafe.sol:65-70 catches any validateErc1271Signature revert and returns INVALID, so the INVALID assertion in scripts/UpgradeSafeImpl.s.sol:71-76 does not prove the linked library ran. Same caveat applies to the OP answer probes at scripts/gnosis-txs/UpgradeSafeImplOP3CP.s.sol:283-290, although the OP script separately guards the recorded library/code and linked bytecode at lines 100, 123, and 129. The foreign EIP-712 invariant is covered by unit test with owner sigs at test/safe/SafeWethAndErc1271.t.sol:142-150; the deploy script empty-signer probe alone does not prove it. Smallest next step: make the deploy health probe call validateErc1271Signature directly for a decodable INVALID case, or add a positive linked-library MAGIC path, instead of using isValidSignature as link proof.

File: src/safe/EtherFiSafe.sol [hunk 3]
```diff
@@ -13,5 +15,102 @@ import { ArrayDeDupLib } from "../libraries/ArrayDeDupLib.sol";
...
+    ///
+    ///      A pause passes ETH through rather than rejecting it: that restores the exact pre-upgrade
+    ///      behaviour, where a paused wrapper cannot strand an inbound transfer. Cheapest checks first —
+    ///      the value and WETH compares have to fit a 2300-gas stipend, `gasleft` is 2 gas, and
+    ///      `inModuleBatch` is a 100-gas tload, so neither the module lookup nor the pause read is
+    ///      reached on the hot paths.
+    receive() external payable override {
+        if (msg.value == 0 || msg.sender == WETH) return;
+
+        if (gasleft() < WRAP_GAS_FLOOR || inModuleBatch() || isModuleEnabled(msg.sender) || dataProvider.paused()) return;
```
[the remark this answers is not in the document]
Comment (ethanai-ethfi) [reply src/safe/EtherFiSafe.sol:111]: Valid on live head a398ba1. receive still returns success when gasleft() < WRAP_GAS_FLOOR, and SafeWethAndErc1271 tests explicitly assert a 30k send stays native plus fuzz all 20k-200k forwarded gas as successful. That gives eth_estimateGas a cheap non-reverting path instead of the WETH deposit path. Smallest next step: make under-floor ordinary sends non-successful, or otherwise force estimates above the wrap floor, while keeping WETH/module/batch exceptions.

**Review by shivam-ef — APPROVED** [review]
